File: src/commands/webhooks.rs
use crate::client::AttioClient;
use crate::error::Result;
use crate::output::{print_many, print_one, OutputFormat};
use crate::types::{DataResponse, ListResponse, Webhook};
use serde::Serialize;

/// List webhooks.
pub async fn list(client: &AttioClient, json: bool) -> Result<()> {
    let response: ListResponse<Webhook> = client.get("/webhooks").await?;
    print_many(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

/// Get a specific webhook.
pub async fn get(client: &AttioClient, webhook_id: &str, json: bool) -> Result<()> {
    let response: DataResponse<Webhook> = client
        .get(&format!("/webhooks/{}", webhook_id))
        .await?;
    print_one(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

/// Create a new webhook.
pub async fn create(
    client: &AttioClient,
    target_url: &str,
    subscriptions: &str,
    json: bool,
) -> Result<()> {
    #[derive(Serialize)]
    struct CreateRequest {
        data: CreateData,
    }

    #[derive(Serialize)]
    struct CreateData {
        target_url: String,
        subscriptions: Vec<Subscription>,
    }

    #[derive(Serialize)]
    struct Subscription {
        event_type: String,
    }

    // Parse comma-separated subscriptions
    let subs: Vec<Subscription> = subscriptions
        .split(',')
        .map(|s| Subscription {
            event_type: s.trim().to_string(),
        })
        .collect();

    let request = CreateRequest {
        data: CreateData {
            target_url: target_url.to_string(),
            subscriptions: subs,
        },
    };

    let response: DataResponse<Webhook> = client.post("/webhooks", &request).await?;
    print_one(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

/// Update a webhook.
pub async fn update(
    client: &AttioClient,
    webhook_id: &str,
    target_url: Option<&str>,
    subscriptions: Option<&str>,
    json: bool,
) -> Result<()> {
    #[derive(Serialize)]
    struct UpdateRequest {
        data: UpdateData,
    }

    #[derive(Serialize)]
    struct UpdateData {
        #[serde(skip_serializing_if = "Option::is_none")]
        target_url: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        subscriptions: Option<Vec<Subscription>>,
    }

    #[derive(Serialize)]
    struct Subscription {
        event_type: String,
    }

    let subs = subscriptions.map(|s| {
        s.split(',')
            .map(|e| Subscription {
                event_type: e.trim().to_string(),
            })
            .collect()
    });

    let request = UpdateRequest {
        data: UpdateData {
            target_url: target_url.map(|s| s.to_string()),
            subscriptions: subs,
        },
    };

    let response: DataResponse<Webhook> = client
        .patch(&format!("/webhooks/{}", webhook_id), &request)
        .await?;
    print_one(&response.data, OutputFormat::from_json_flag(json))?;
    Ok(())
}

File: src/attio/commands/webhooks.py
"""Webhook commands."""

import click

from attio.columns import WEBHOOK_COLUMNS
from attio.commands.common import get_client
from attio.output import output_many, output_one


@click.group()
def webhooks() -> None:
    """Manage webhooks."""


@webhooks.command("list")
@click.option("--json", "as_json", is_flag=True, help="Output as JSONL")
def webhooks_list(as_json: bool) -> None:
    """List webhooks."""
    with get_client() as client:
        response = client.get("/webhooks")
        output_many(response["data"], WEBHOOK_COLUMNS, as_json)


@webhooks.command("retrieve")
@click.argument("webhook_id")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def webhooks_retrieve(webhook_id: str, as_json: bool) -> None:
    """Retrieve webhook details."""
    with get_client() as client:
        response = client.get(f"/webhooks/{webhook_id}")
        output_one(response["data"], WEBHOOK_COLUMNS, as_json)


@webhooks.command("create")
@click.option("--target-url", required=True, help="Target URL")
@click.option("--subscriptions", required=True, help="Comma-separated event types")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def webhooks_create(target_url: str, subscriptions: str, as_json: bool) -> None:
    """Create a new webhook."""
    subs = [{"event_type": s.strip()} for s in subscriptions.split(",")]
    with get_client() as client:
        response = client.post(
            "/webhooks",
            {"data": {"target_url": target_url, "subscriptions": subs}},
        )
        output_one(response["data"], WEBHOOK_COLUMNS, as_json)


@webhooks.command("update")
@click.argument("webhook_id")
@click.option("--target-url", help="New target URL")
@click.option("--subscriptions", help="New subscriptions (comma-separated)")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def webhooks_update(webhook_id: str, target_url: str, subscriptions: str, as_json: bool) -> None:
    """Update a webhook."""
    data = {}
    if target_url:
        data["target_url"] = target_url
    if subscriptions:
        data["subscriptions"] = [{"event_type": s.strip()} for s in subscriptions.split(",")]

    with get_client() as client:
        response = client.patch(f"/webhooks/{webhook_id}", {"data": data})
        output_one(response["data"], WEBHOOK_COLUMNS, as_json)
